lib/src/migrators/module/references.dart

```diff
@@ -0,0 +1,326 @@
+          .union(getFunctionReferences.keysForValue(declaration));
+    } else if (declaration is MixinRule) {
+      references = mixins.keysForValue(declaration);
+    } else {
+      references = variables.keysForValue(declaration);
+    }
+    return references.any(
+        (reference) => reference.span.sourceUrl != declaration.span.sourceUrl);
+  }
+
+  /// Finds the original declaration of the variable referenced in [reference].
+  SassNode /*VariableDeclaration|Argument*/ originalDeclaration(
+      VariableExpression reference) {
+    var declaration = variables[reference];
+    return variableReassignments[declaration] ?? declaration;
+  }
+
+  /// For a given declaration or reference, returns the set of all nodes
+  /// associated with it, including itself.
+  ///
+  /// For mixins, this is the declaring MixinRule and all referencing
+  /// IncludeRules.
+  ///
+  /// For functions, this is the declaring FunctionRule and all referencing
+  /// FunctionRules (both regular and statically-determined get-function calls).
+  ///
+  /// For variables, this is the original declaration, all reassignments, and
+  /// all referencing VariableExpressions.
+  Set<SassNode> allNodes(SassNode node) {
+    if (node is IncludeRule) return allNodes(mixins[node]);
+    if (node is FunctionExpression) return allNodes(functions[node]);
+    if (node is VariableExpression) return allNodes(originalDeclaration(node));
+    if (node is MixinRule) return {node, ...mixins.keysForValue(node)};
+    if (node is FunctionRule) {
+      return {
+        node,
+        ...functions.keysForValue(node),
+        ...getFunctionReferences.keysForValue(node)
+      };
+    }
+    return {
+      node,
+      ...variables.keysForValue(node),
+      for (var reassignment in variableReassignments.keysForValue(node))
+        ...variables.keysForValue(reassignment)
+    };
+  }
+
+  /// Constructs a new References object based on the stylesheet at [entrypoint]
+  /// and its dependencies.
+  static References build(Uri entrypoint) =>
+      _ReferenceVisitor().build(entrypoint);
+}
+
+/// A visitor that builds a References object.
+class _ReferenceVisitor extends RecursiveAstVisitor {
+  /// The References object being built.
+  final _references = References._();
+
+  /// The current global scope.
+  ///
+  /// This persists across imports, but not across module uses.
+  Scope _scope;
+
+  /// Mapping from canonical stylesheet URLs to the global scope of the module
+  /// contained within it.
+  ///
+  /// Note: Stylesheets only depended on through imports will not have their
+  /// own scope in this map; they will instead share a global scope with the
+  /// stylesheet that imported them.
+  final _moduleScopes = <Uri, Scope>{};
+
+  /// Namespaces present within the current stylesheet.
+  ///
+  /// Note: Unlike the similar property in _ModuleMigrationVisitor, this only
+  /// includes namespaces for use rules that already exist within the file.
```

Comment:
```suggestion
  /// includes namespaces for `@use` rules that already exist within the file.
```

It's generally good to explicitly refer to at-rules with their full names in code font, to help make it visually clear that you're not just using the corresponding English word. This is especially useful for non-native speakers.